### PR TITLE
Allow batch reading of mesh chunks

### LINKED ISSUE #6752: Optimize loading of precomputed meshes of oversegmentation
If a meshfile exists for the oversegmentation, webknossos can load meshes for any agglomerate mapping, by fetching the oversegmentation meshes for all segments of an agglomerate and giving them all the same color. The back-end side for this is implemented in `listMeshChunksForSegmentForVersion` and `readMeshChunkForVersion` in `DSMeshController.scala`.

We had a couple of ideas for optimizing the performance
 - the front-end currently sends one request per chunk per segment id
 - instead it could send just one request per chunk, including a list of meshFile/byteOffset/byteSize tuples for all relevant segments, reducing the number of http requests
 - the back-end could sort this by byte offset to optimize for spinning-disk based reading of the hdf5 file
 - optional: similar to bucket loading, we can think about loading more than one chunk per http request

Proposal: request contains json list, as described above. Reply is binary, consisting of
 - Length of jumptable (one long)
 - The jumptable (start offsets for each segment content)
 - the concatenated draco contents

### PR DESCRIPTION
**Pull request by frcroth**
### URL of deployed dev instance (used for testing):
- https://optimizepremeshesoverseg.webknossos.xyz/

### Steps to test:
- open l4dense_motta_et_al_dev_v2 in view mode
- enable mapping 80
- load a precomputed mesh for a segment (e.g., at 2905, 4566, 1786)

### TODO
- [X] Adapt front end
- [X] Use new format on other route
- [X] Tweak performance so that it generalizes over different mesh files
- ~[ ] Use uint32 instead of uint64 for jump table so that Uint32Array can be used in JS~
- [x] ensure proper error handling when decoding fails (independent of this PR, strictly speaking)
- [ ] debug why some requests cannot be decoded (independent of this PR, strictly speaking)

### Issues:
- fixes #6752 

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Removed dev-only changes like prints and application.conf edits
- [ ] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [ ] Needs datastore update after deployment

### CODE REVIEW

**Comment by frcroth** [comment]
@philippotto Backend should be ready (though not completely tested)

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/MeshFileService.scala
Comment (philippotto) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/MeshFileService.scala:416]: @frcroth I adapted the jump table so that its entries are in byte and not in elements (in other words, I added the `8 * ...` part in line 415). Probably the "magic" 8 can be expressed by retrieving the byte count of a `Long`? Could you tidy this up?
Also, I have another wish: Could you adapt the jump table so that its elements are not Longs, but instead unsigned 32-bit integers. 64-bit numbers are a bit tedious to handle in JS and since we won't request 4 GB mesh data in one request, 32-bit should be enough, I assume.

**Comment by philippotto** [comment]
I just noticed that `THREE.DRACOLoader: Unexpected geometry type.` is thrown on the dev instance for some chunk (most of the mesh loads fine). This is already the case on master, too. See the testing steps in the PR description which trigger this.

The front-end doesn't handle this correctly currently (the spinner just keeps spinning). I'll add proper error handling and then we should be able to debug this further.

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/MeshFileService.scala
Comment (philippotto) [reply webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/MeshFileService.scala:416]: @frcroth nevermind! the front-end can compute the jump table on its own, because it already knows the byteSizes of the individual chunks. I'll remove the jump table code from the back-end next week.

File: frontend/javascripts/libs/DRACOLoader.ts
Comment (philippotto) [inline frontend/javascripts/libs/DRACOLoader.ts:3]: copied from threejs

**Comment by philippotto** [comment]
I think, the PR is ready for review. @daniel-wer It can make sense to review the PR by commit range (I did some refactoring of the isosurface saga were I simply extracted/moved some code). Also, I upgraded the draco libs.

@frcroth @fm3 See my previous comment regarding an error where a mesh chunk cannot be decoded. I added better error handling and reporting, but I couldn't really figure out what's happening. I only found one segment where one mesh request fails and that is:

```
Requested chunk: {position: [1024, 5120, 1024], byteOffset: 3920109208, byteSize: 1818}
Error: {type: 'error', id: 9592, error: 'THREE.DRACOLoader: Unexpected geometry type.'}
```

It happens with `l4dense_motta_et_al_dev_v2/segmentation/meshes/meshfile_mapping0.hdf5` with mesh id 355 and mapping 80. The mesh file is 13 GB big..

Maybe you have an idea what could be causing this. However, I think, it's not blocking this PR as this is already an issue on master.

**Comment by philippotto** [comment]
Bug can be seen isolated on this dev instance: https://provokemeshdecodingerror.webknossos.xyz/datasets/sample_organization/l4dense_motta_et_al_dev_v2/view#2816,4353,1792,0,1.3 (enable mapping 80 and load a mesh).

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/MeshFileService.scala
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/MeshFileService.scala:408]: Bare assertions are hard to identify when something goes wrong in production. Please add an error message with `?~>`

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/MeshFileService.scala
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/MeshFileService.scala:409]:
```suggestion
      encoding <- data.map(d => d._2).headOption
```
(will create a Fox.failure rather than an exception if something goes wrong)

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/MeshFileService.scala
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/MeshFileService.scala:383]: Could you test if Fox.serialCombined slows this down significantly on your machine? The current implementation with Fox.combined and map may parallelize over all entries, which could actually be slower on a spinning disk.

**Review by fm3 — COMMENTED** [review]
Works for me :) Great speed increase! I added a note on parallelization.

@philippotto could you write an issue for the draco decoding problem?

**Comment by philippotto** [comment]
> @philippotto could you write an issue for the draco decoding problem?

Done! https://github.com/scalableminds/webknossos/issues/7044

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/MeshFileService.scala
Comment (fm3) [reply webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/MeshFileService.scala:383]: Thanks for looking into this! Yes, I’d say for the moment we should optimize for spinning disk (which is also what the sorting-by-byte-offset step is for)

File: CHANGELOG.unreleased.md
Comment (daniel-wer) [inline CHANGELOG.unreleased.md:17]:
```suggestion
- Loading of precomputed meshes got significantly faster (especially when using a mesh file for an oversegmentation with an applied agglomerate mapping). [#7001](https://github.com/scalableminds/webknossos/pull/7001)
```

File: frontend/javascripts/oxalis/model/sagas/isosurface_saga.ts
Comment (daniel-wer) [inline frontend/javascripts/oxalis/model/sagas/isosurface_saga.ts:666]: Can likely be removed before merging

**Review by daniel-wer — APPROVED** [review]
Works very well, awesome speed boost!

File: public/wasm/draco_decoder.wasm
Comment (philippotto) [reply public/wasm/draco_decoder.wasm:1]: No, it was checked in and used :) However, its size increased from 275 KB to 280 KB which is why github puts the "+4.17 KB (100%)" label. I was confused by this, too, since the 100% suggests that it's completely new. However, it probably means that 100% of the changes are additions? :sweat_smile:

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/MeshFileService.scala
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/MeshFileService.scala:384]: could you add a comment that the sorting is done to optimize for spinning disk access?

File: public/wasm/draco_decoder.wasm
Comment (daniel-wer) [reply public/wasm/draco_decoder.wasm:1]: Ah, yes that's slightly confusing :sweat_smile:

**Comment by philippotto** [comment]
@frcroth just a small ping in case this PR slipped through your inbox :)

**Comment by frcroth** [comment]
> @frcroth just a small ping in case this PR slipped through your inbox :)

@philippotto I'm a bit out of the loop on this one. What should I do?

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/MeshFileService.scala
Comment (frcroth) [reply webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/MeshFileService.scala:405]: Don't think anything can go wrong because of the statument in the line above, but I guess it is better style

**Comment by philippotto** [comment]
> > @frcroth just a small ping in case this PR slipped through your inbox :)
> 
> @philippotto I'm a bit out of the loop on this one. What should I do?

Only the PR feedback was missing, but I guess this is good now :+1: @fm3 Do you want to leave a final approval?

**Review by fm3 — APPROVED** [review]
Boom :rocket:

**Pull request merged**
